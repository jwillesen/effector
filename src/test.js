import {Effector} from './effector'
import {createCallEffect} from './utils'

export class TestEffectRegistry {
  static theInstance = null

  static get instance () {
    if (this.theInstance == null) this.theInstance = new TestEffectRegistry()
    return this.theInstance
  }

  static set instance (newInstance) {
    this.theInstance = newInstance
  }

  mockedCalls = [] // array of objects: {effect: <a call effect>, mockFn: <provided mock fn>}
  mockCall (toInvoke, mockFn) {
    const effect = createCallEffect(toInvoke)
    this.mockedCalls.push({effect, mockFn})
    return mockFn
  }

  call (effect) {
    const mockedCall = this.findMockCallForEffect(effect)
    if (mockedCall == null) throw new Error('Tried to invoke an unmocked call')
    if (typeof mockedCall.mockFn === 'function') {
      return mockedCall.mockFn(...effect.args)
    } else {
      return mockedCall.mockFn
    }
  }

  reset () {
    this.mockedCalls = []
  }

  findMockCallForEffect (effect) {
    return this.mockedCalls.find(mockedCall => this.equalEffects(effect, mockedCall.effect))
  }

  equalEffects (first, second) {
    return first.obj === second.obj && first.fn === second.fn
  }
}

export function createTestMiddleware (testEffectRegistry) {
  return function (effect, next) {
    return testEffectRegistry.call(effect)
    // purposefully not calling next because we don't want to actually invoke
    // these calls while testing. The provided mock function can invoke it if
    // desired.
  }
}

export function mockCall (invoked, fn) {
  return TestEffectRegistry.instance.mockCall(invoked, fn)
}

export function useMockCalls () {
  Effector.instance = new Effector([
    createTestMiddleware(TestEffectRegistry.instance),
  ])
}

export function resetMockCalls () {
  TestEffectRegistry.instance.reset()
}

export const effector = {
  mockCall,
  useMockCalls,
  resetMockCalls,
}

export default effector
