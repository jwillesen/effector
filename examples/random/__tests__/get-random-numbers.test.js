import {mockCall, useMockCalls, resetMockCalls} from 'effector/lib/test'
import getRandomNumbers from '../get-random-numbers'
import axios from 'axios'

useMockCalls()
afterEach(() => {
  resetMockCalls()
})

it('fetches random numbers', async () => {
  // mockCall invokes the passed function when a matching call is made.
  mockCall([axios, 'get'], (url, params) => {
    expect(url).toMatch(/random.org/)
    expect(params).toMatchSnapshot()
    return {data: 'result-data'}
  })
  // mockCall returns its second parameter as a convenience for creating mock functions.
  const mockLog = mockCall(console.log, jest.fn())
  expect(await getRandomNumbers()).toBe('result-data')
  expect(mockLog).toHaveBeenCalledWith('result-data')
})
