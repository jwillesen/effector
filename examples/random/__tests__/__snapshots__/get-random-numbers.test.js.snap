// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches random numbers 1`] = `
Object {
  "params": Object {
    "base": 10,
    "col": 10,
    "format": "plain",
    "max": 100,
    "min": 1,
    "num": 100,
    "rnd": "new",
  },
}
`;
